text/0000-type-macros.md

```diff
@@ -0,0 +1,467 @@
+
+```rust
+let xs: Cons<&str, Cons<bool, Cons<Vec<u64>, Nil>>> = hlist!["foo", false, vec![0u64]];
+```
+
+Allowing type macros as this RFC proposes would allows us to be
+able to use Rust's macros to improve writing the HList type as
+well. The complete example follows:
+
+```rust
+// term-level macro for HLists
+macro_rules! hlist {
+    {} => { Nil };
+    {=> $($elem:tt),+ } => { hlist_pat!($($elem),+) };
+    { $head:expr, $($tail:expr),* } => { Cons($head, hlist!($($tail),*)) };
+    { $head:expr } => { Cons($head, Nil) };
+}
+
+// term-level HLists in patterns
+macro_rules! hlist_pat {
+    {} => { Nil };
+    { $head:pat, $($tail:tt),* } => { Cons($head, hlist_pat!($($tail),*)) };
+    { $head:pat } => { Cons($head, Nil) };
+}
+
+// type-level macro for HLists
+macro_rules! HList {
+    {} => { Nil };
+    { $head:ty } => { Cons<$head, Nil> };
+    { $head:ty, $($tail:ty),* } => { Cons<$head, HList!($($tail),*)> };
+}
+
+let xs: HList![&str, bool, Vec<u64>] = hlist!["foo", false, vec![0u64]];
+```
+
+Operations on HLists can be defined by recursion, using traits with
+associated type outputs at the type-level and implementation methods
+at the term-level.
+
+The HList append operation is provided as an example. type macros are
+used to make writing append at the type level (see `Expr!`) more
+convenient than specifying the associated type projection manually:
+
+```rust
+use std::ops;
+
+// nil case for HList append
+impl<Ys: HList> ops::Add<Ys> for Nil {
+    type Output = Ys;
+
+    fn add(self, rhs: Ys) -> Ys {
+        rhs
+    }
+}
+
+// cons case for HList append
+impl<Rec: HList + Sized, X, Xs: HList, Ys: HList> ops::Add<Ys> for Cons<X, Xs> where
+    Xs: ops::Add<Ys, Output = Rec>,
+{
+    type Output = Cons<X, Rec>;
+
+    fn add(self, rhs: Ys) -> Cons<X, Rec> {
+        Cons(self.0, self.1 + rhs)
+    }
+}
+
+// type macro Expr allows us to expand the + operator appropriately
+macro_rules! Expr {
+    { ( $($LHS:tt)+ ) } => { Expr!($($LHS)+) };
+    { HList ! [ $($LHS:tt)* ] + $($RHS:tt)+ } => { <Expr!(HList![$($LHS)*]) as std::ops::Add<Expr!($($RHS)+)>>::Output };
+    { $LHS:tt + $($RHS:tt)+ } => { <Expr!($LHS) as std::ops::Add<Expr!($($RHS)+)>>::Output };
+    { $LHS:ty } => { $LHS };
+}
+
+// test demonstrating term level `xs + ys` and type level `Expr!(Xs + Ys)`
+#[test]
+fn test_append() {
+    fn aux<Xs: HList, Ys: HList>(xs: Xs, ys: Ys) -> Expr!(Xs + Ys) where
+        Xs: ops::Add<Ys>
+    {
+        xs + ys
+    }
+    let xs: HList![&str, bool, Vec<u64>] = hlist!["foo", false, vec![]];
+    let ys: HList![u64, [u8; 3], ()] = hlist![0, [0, 1, 2], ()];
+
+    // demonstrate recursive expansion of Expr!
+    let zs: Expr!((HList![&str] + HList![bool] + HList![Vec<u64>]) +
+                  (HList![u64] + HList![[u8; 3], ()]) +
+                  HList![])
+        = aux(xs, ys);
+    assert_eq!(zs, hlist!["foo", false, vec![], 0, [0, 1, 2], ()])
+}
+```
+
+### Additional Examples ###
```

Comment:
I think this entire section should go, since while interesting it doesn't really further the RFCs point.